### PR TITLE
Introduction of NB FS Native module

### PR DESCRIPTION
**Pull request by jeniawhite**
### Explain the changes
1. Addition of FS native layer
2. Changing the namespace_fs to use nb fs native module
3. Added FileWrapper, DirWrapper
4. Ops added: stat, unlink, rename, rmkir, mkdir, writeFile, readFile, readdir

GAPS (#TODO):
1. Add UID capability
2. Rewrite createReadStream and createWriteStream
3. Fill up unit_tests for native
4. Passing more complex options to ops (bufferSize and etc...)

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

### CODE REVIEW

File: src/deploy/NVA_build/builder.Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM registry.centos.org/centos:8 
+FROM centos:8
```
Comment (guymguym) [inline src/deploy/NVA_build/builder.Dockerfile:1]: @jeniawhite Is this change intended? CC @liranmauda

Comment (jeniawhite) [reply src/deploy/NVA_build/builder.Dockerfile:1]: My mistake, the repo did not work for me. Disregard.
Thank you.

File: src/native/fs/fs_napi.cpp [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+    {
+        LOG("FS::FSWorker::OnOK: undefined " << _desc);
+        _deferred.Resolve(Env().Undefined());
+    }
+    virtual void OnError(Napi::Error const &error)
+    {
+        Napi::Env env = Env();
+        LOG("FS::FSWorker::OnError: " << _desc << " " << DVAL(error.Message()));
+        auto obj = error.Value();
+        obj.Set("code", Napi::String::New(env, uv_err_name(uv_translate_sys_error(_errno))));
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:70]: Maybe need to set this only `if (_errno)`

File: src/native/fs/fs_napi.cpp [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+        int fd = open(_path.c_str(), O_WRONLY | O_CREAT);
+        if (fd < 0) {
+            SetSyscallError();
+            return;
+        }
+
+        ssize_t len = write(fd, _data, _len);
+        if (len < 0) {
+            SetSyscallError();
+            return;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:219]: I think we want to avoid return here to try and close the fd on error

File: src/native/fs/fs_napi.cpp [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+        }
+        if ((size_t)len != _len) {
+            SetError(XSTR() << "Writefile: partial write error " << DVAL(len) << DVAL(_len));
+            return;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:223]: again - best continue after setting the error and try to close the fd

File: src/native/fs/fs_napi.cpp [hunk 5]
```diff
@@ -0,0 +1,747 @@
...
+    {
+    }
+    void Begin(std::string desc)
+    {
+        _desc = desc;
+        LOG("FS::FSWorker::Begin: " << _desc);
+    }
+    void SetSyscallError()
+    {
+        _errno = errno;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:56]: I think we want to only set it if _errno is not set (0) otherwise log and return.

File: src/native/fs/fs_napi.cpp [hunk 6]
```diff
@@ -0,0 +1,747 @@
...
+            SetSyscallError();
+            return;
+        }
+
+        ssize_t len = write(fd, _data, _len);
+        if (len < 0) {
+            SetSyscallError();
+            return;
+        }
+        if ((size_t)len != _len) {
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:221]: else if iss needed if we remove the previous return

File: src/native/fs/fs_napi.cpp [hunk 7]
```diff
@@ -0,0 +1,747 @@
...
+        _len = stat_res.st_size;
+        _data = new uint8_t[_len];
+
+        uint8_t *p = _data;
+        int remain = _len;
+        while (remain > 0) {
+            ssize_t len = read(fd, p, remain);
+            if (len < 0) {
+                SetSyscallError();
+                return;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:278]: again, best try to break to close the fd

File: src/native/fs/fs_napi.cpp [hunk 8]
```diff
@@ -0,0 +1,747 @@
...
+        if (fd < 0) {
+            SetSyscallError();
+            return;
+        }
+    
+        struct stat stat_res;
+        int r = fstat(fd, &stat_res);
+        if (r != 0) {
+            SetSyscallError();
+            return;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:266]: need to close here

File: src/native/fs/fs_napi.cpp [hunk 9]
```diff
@@ -0,0 +1,747 @@
...
+
+
+/**
+ * Readdir is an fs op
+ */
+struct Readdir : public FSWorker
+{
+    std::string _path;
+    bool _withFileTypes;
+    DIR *_dir;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:302]: _dir should not be a member here, move to local in Execute

File: src/native/fs/fs_napi.cpp [hunk 10]
```diff
@@ -0,0 +1,747 @@
...
+        Napi::Env env = Env();
+        Napi::Array res = Napi::Array::New(env);
+        int index = 0;
+        if (_withFileTypes) {
+            for (auto it = _entries.begin(); it != _entries.end(); ++it) {
+                // Ingnore parent and current directories 
+                if (it->name == "." || it->name == "..") {
+                    continue;
+                }
+                auto dir_rec = Napi::Array:: Object::New(env);
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:356]: something here looks wrong. need to be Napi::Object::New.

File: src/native/fs/fs_napi.cpp [hunk 11]
```diff
@@ -0,0 +1,747 @@
...
+            }
+        }
+
+        int r = closedir(_dir);
+        if (r) SetSyscallError();
+    }
+    virtual void OnOK()
+    {
+        Napi::Env env = Env();
+        Napi::Array res = Napi::Array::New(env);
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:348]: Best init with size `_entries.size() - 2` (without `.` and `..`)

File: src/native/fs/fs_napi.cpp [hunk 12]
```diff
@@ -0,0 +1,747 @@
...
+
+//         // TODO - read(fd, buf...)
+//     }
+// };
+
+
+// /**
+//  * Writefile is an fs op
+//  */
+// struct Writefile : public FSWorker
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:527]: still needed? I see we already have the Writefile op above

File: src/native/fs/fs_napi.cpp [hunk 13]
```diff
@@ -0,0 +1,747 @@
...
+        }
+    }
+    virtual void OnOK()
+    {
+        Napi::Env env = Env();
+        _deferred.Resolve(Napi::Number::New(env, _br));
+    }
+};
+
+// struct FileWrite : public FSWorker
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:498]: Are you going to complete this op in this PR?

File: src/test/unit_tests/test_nb_native_fs.js [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+// const _ = require('lodash');
+const fs = require('fs');
+const mocha = require('mocha');
+const assert = require('assert');
+const nb_native = require('../../util/nb_native');
+
+mocha.describe('nb_native fs', function() {
+
+    // mocha.describe('stat', function() {
```
Comment (guymguym) [inline src/test/unit_tests/test_nb_native_fs.js:12]: what's up with this test? Why is it all commented out?

File: src/sdk/namespace_fs.js [hunk 15]
```diff
@@ -711,14 +724,13 @@ class NamespaceFS {
         // IMPORTANT NOTICE - we must return an etag that contains a dash!
         // because this is the criteria of S3 SDK to decide if etag represents md5
         // and perform md5 validation of the data.
-        const ident_str = 'inode-' + stat.ino.toString() + '-mtime-' + stat.mtimeMs.toString();
+        const ident_str = 'inode-' + stat.ino.toString() + '-mtime-' + stat.mtime.toString();
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:727]: what is the resolution of the mtime we return from stat?

File: src/sdk/namespace_fs.js [hunk 16]
```diff
@@ -783,13 +793,60 @@ class NamespaceFS {
...
             throw err;
         }
     }
 
+    async _create_path(dir) {
+        const sep = path.sep;
+        const initDir = path.isAbsolute(dir) ? sep : '';
+        const baseDir = '.';
+
+        return dir.split(sep).reduce(async (parentDir, childDir) => {
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:809]: is this code copied from somewhere?
I would not use this without permission to copy.
In any case I don't like how this is written at all.
I would much prefer to write like this:
```js
let dir_path = path.isAbsolute(dir) ? path.sep : '';
for (const item of dir.split(path.sep)) {
    dir_path = path.join(dir_path, item);
    try {
        await nb_native().mkdir(dir_path);
    } catch (err) {
        if (err.code !== 'EEXIST') throw err;
    }
}
```

File: src/sdk/namespace_fs.js [hunk 17]
```diff
@@ -783,13 +793,60 @@ class NamespaceFS {
...
+        let entries = await nb_native().readdir(dir, /* withFileTypes */ true);
+        let results = await Promise.all(entries.map(entry => {
+          let fullPath = path.join(dir, entry.name);
+        //   if (fullPath.startsWith('/tmp/')) {
+          let task = isDirectory(entry) ? this._folder_delete(fullPath) : nb_native().unlink(fullPath);
+          return task.catch(error => ({ error }));
+        //   }
+        }));
+        results.forEach(result => {
+          // TODO: Handle errors correctly
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:841]: seems like this code is still work in progress - is it ready for review?

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/sdk/namespace_fs.js (hunk 15 above)
Comment (jeniawhite) [reply src/sdk/namespace_fs.js:727]: The timestamp is in seconds

Comment (jeniawhite) [reply src/sdk/namespace_fs.js:727]: #if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
	struct  timespec st_atimespec;  /* time of last access */
	struct  timespec st_mtimespec;  /* time of last data modification */
	struct  timespec st_ctimespec;  /* time of last status change */
#else
	time_t          st_atime;       /* [XSI] Time of last access */
	long            st_atimensec;   /* nsec of last access */
	time_t          st_mtime;       /* [XSI] Last data modification time */
	long            st_mtimensec;   /* last data modification nsec */
	time_t          st_ctime;       /* [XSI] Time of last status change */
	long            st_ctimensec;   /* nsec of last status change */
#endif

As you can see I can take the value from `st_mtimensec` but not on darwin, so I will do an if there.

Comment (jeniawhite) [reply src/sdk/namespace_fs.js:727]: `
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
	struct  timespec st_atimespec;  /* time of last access */
	struct  timespec st_mtimespec;  /* time of last data modification */
	struct  timespec st_ctimespec;  /* time of last status change */
#else
	time_t          st_atime;       /* [XSI] Time of last access */
	long            st_atimensec;   /* nsec of last access */
	time_t          st_mtime;       /* [XSI] Last data modification time */
	long            st_mtimensec;   /* last data modification nsec */
	time_t          st_ctime;       /* [XSI] Time of last status change */
	long            st_ctimensec;   /* nsec of last status change */
#endif
`
And this is the alias
`
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
#define st_atime st_atimespec.tv_sec
#define st_mtime st_mtimespec.tv_sec
#define st_ctime st_ctimespec.tv_sec
#define st_birthtime st_birthtimespec.tv_sec
#endif /* (!_POSIX_C_SOURCE || _DARWIN_C_SOURCE) */
`
As you can see on darwin we do not have the option for nanoseconds, but I will add a code that will pull from nanoseconds as well and on darwin, I will just multiply for our local tests to pass.

Comment (guymguym) [reply src/sdk/namespace_fs.js:727]: Darwin uses `struct timespec` which provides nano seconds too:
```
$ tail /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/_types/_timespec.h
#define _STRUCT_TIMESPEC        struct timespec

#include <machine/types.h> /* __darwin_time_t */

_STRUCT_TIMESPEC
{
	__darwin_time_t tv_sec;
	long            tv_nsec;
};
#endif /* _STRUCT_TIMESPEC */
```

Comment (guymguym) [reply src/sdk/namespace_fs.js:727]: In any case - returning seconds is wrong - we need to convert to ms in order to be compatible with js dates.

File: src/native/fs/fs_napi.cpp [hunk 18]
```diff
@@ -0,0 +1,747 @@
...
+        res["nlink"] = Napi::Number::New(env, _stat_res.st_nlink);
+        res["uid"] = Napi::Number::New(env, _stat_res.st_uid);
+        res["gid"] = Napi::Number::New(env, _stat_res.st_gid);
+        res["rdev"] = Napi::Number::New(env, _stat_res.st_rdev);
+        res["size"] = Napi::Number::New(env, _stat_res.st_size);
+        res["blksize"] = Napi::Number::New(env, _stat_res.st_blksize);
+        res["blocks"] = Napi::Number::New(env, _stat_res.st_blocks);
+        res["atime"] = Napi::Number::New(env, _stat_res.st_atime);
+        res["mtime"] = Napi::Number::New(env, _stat_res.st_mtime);
+        res["ctime"] = Napi::Number::New(env, _stat_res.st_ctime);
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:110]: Based on what we discussed below on times - we need to convert the times to milliseconds to be compatible with js dates. Also need to take into account the nano seconds precision which requires some difference between Mac and Linux. This example works on both linux and mac:

```
$ cat a.cpp
#include <stdio.h>
#include <sys/stat.h>
int main()
{
	struct stat st;
	int r = stat(".", &st);
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
	double d = (double(1e3) * st.st_mtimespec.tv_sec) + (double(1e-6) * st.st_mtimespec.tv_nsec);
#else
	double d = (double(1e3) * st.st_mtim.tv_sec) + (double(1e-6) * st.st_mtim.tv_nsec);
#endif
	printf("mtime = %lf \n", d);
	return 0;
}
```

File: src/native/fs/fs_napi.cpp (hunk 13 above)
Comment (jeniawhite) [reply src/native/fs/fs_napi.cpp:515]: Next PR, I will also probably rewrite the logic for uploads and reads to use the FileWrapper

File: src/test/unit_tests/test_nb_native_fs.js (hunk 14 above)
Comment (jeniawhite) [reply src/test/unit_tests/test_nb_native_fs.js]: I will fill this up in the next PR as well

File: src/sdk/namespace_fs.js [hunk 19]
```diff
@@ -75,8 +80,8 @@ const dir_cache = new LRUCache({
         return { time, stat, sorted_entries, usage };
     },
     validate: async ({ stat }, dir_path) => {
-        const new_stat = await fs.promises.stat(dir_path);
-        return (new_stat.ino === stat.ino && new_stat.mtime.getTime() === stat.mtime.getTime());
+        const new_stat = await nb_native().stat(dir_path);
+        return (new_stat.ino === stat.ino && new_stat.mtime === stat.mtime);
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:84]: CAREFUL HERE - this change is wrong now that you changed stat.mtime to be a Date object !
This was a bug I fixed in https://github.com/noobaa/noobaa-core/pull/6394
It would be best to actually compare mtimeMs instead here.
Please fix.

Comment (jeniawhite) [reply src/sdk/namespace_fs.js:84]: Somehow missed leaving the getTime().
I've changed to `mtimeMs`.
Thanks.

File: src/sdk/namespace_fs.js [hunk 20]
```diff
@@ -62,11 +67,11 @@ const dir_cache = new LRUCache({
     name: 'nsfs-dir-cache',
     load: async dir_path => {
         const time = Date.now();
-        const stat = await fs.promises.stat(dir_path);
+        const stat = await nb_native().stat(dir_path);
         let sorted_entries;
         let usage = config.NSFS_DIR_CACHE_MIN_DIR_SIZE;
         if (stat.size <= config.NSFS_DIR_CACHE_MAX_DIR_SIZE) {
-            sorted_entries = await fs.promises.readdir(dir_path, { withFileTypes: true });
+            sorted_entries = await nb_native().readdir(dir_path, /* withFileTypes */ true);
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:74]: In next PR I recommend to keep the second arg an object like fs.promises does.

Comment (guymguym) [reply src/sdk/namespace_fs.js:74]: Hmm on second thought - do we even need this option??
Isn't it always better to return the file types instead of having this option?

Comment (jeniawhite) [reply src/sdk/namespace_fs.js:74]: I don't see a reason, since this just creates more checks in the native C++ layer and we aren't actually getting any value from it in the NodeJS code.
Do you want me to remove it?

Comment (guymguym) [reply src/sdk/namespace_fs.js:74]: I think so. If we later decide we need it we can always introduce it.

Comment (jeniawhite) [reply src/sdk/namespace_fs.js]: done

**Review by guymguym — APPROVED** [review]
